include/umf/memspace.h

```diff
@@ -0,0 +1,57 @@
+/*
+ *
+ * Copyright (C) 2023 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ */
+
+#ifndef UMF_MEMSPACE_H
+#define UMF_MEMSPACE_H 1
+
+#include <umf/base.h>
+#include <umf/memory_pool.h>
+#include <umf/memory_provider.h>
+#include <umf/memspace_policy.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct umf_memspace_t *umf_memspace_handle_t;
+
+///
+/// \brief Creates new memory pool from memspace and policy.
+/// \param hMemspace handle to memspace
+/// \param hPolicy handle to policy
+/// \param hPool [out] handle to the newly created memory pool
+/// \return UMF_RESULT_SUCCESS on success or appropriate error code on failure.
+///
+umf_result_t umfPoolCreateFromMemspace(umf_memspace_handle_t hMemspace,
+                                       umf_memspace_policy_handle_t hPolicy,
+                                       umf_memory_pool_handle_t *hPool);
+
+///
+/// \brief Creates new memory provider from memspace and policy.
+/// \param hMemspace handle to memspace
+/// \param hPolicy handle to policy
+/// \param hProvider [out] handle to the newly created memory provider
+/// \return UMF_RESULT_SUCCESS on success or appropriate error code on failure.
+///
+umf_result_t
+umfMemoryProviderCreateFromMemspace(umf_memspace_handle_t hMemspace,
+                                    umf_memspace_policy_handle_t hPolicy,
+                                    umf_memory_provider_handle_t *hProvider);
+
+///
+/// \brief Destroys memspace
+/// \param hMemspace handle to memspace
+///
+void umfMemspaceDestroy(umf_memspace_handle_t hMemspace);
```

Comment:
Where is the corresponding create function?

Another question: if we consider only predefined memspaces as a first step do we need Create/Destroy functions for memspaces in a public API?

Reply:
Right now we only have `umfMemspaceCreateFromNumaArray` that is defined in memspace_numa.h

We could move that function to memspace.h for now but in the long term, I think those create functions should be in separate headers (since we might want to have the memspaces in separate libs).

Reply:
> Another question: if we consider only predefined memspaces as a first step do we need Create/Destroy functions for memspaces in a public API?

Ok, good point. @bratpiorka for openMP, we won't need `umfMemspaceCreateFromNumaArray` right? If that's the case I can make it internal for now.

Reply:
Done. I've made those functions private for now.

Reply:
This PR does not expose the predefined memspaces but this will be added in next PR (it requires integrating with hwloc).